### ISSUE TITLE
Interpolation tracking issue

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#50341]
This issue is a roll-up of interpolation-related discussion, requests and issues.

**Function requests:**

- [ ] np.interp, https://github.com/pytorch/pytorch/issues/50334
- [ ] scipy.interpolate.interp1d, https://github.com/pytorch/pytorch/issues/50335
- [ ] scipy.interpolate.InterpolatedUnivariateSpline, https://github.com/pytorch/pytorch/issues/50340
- [ ] scipy.interpolate.RectBivariateSpline, https://github.com/pytorch/pytorch/issues/50337
- [ ] scipy.interpolate.RegularGridInterpolator, https://github.com/pytorch/pytorch/issues/50338
- [ ] scipy.interpolate.griddata, https://github.com/pytorch/pytorch/issues/50339
- [ ] scipy.ndimage.zoom, https://github.com/pytorch/pytorch/issues/50336

**Additional mode requests:**

- [ ] add half_pixel_centers to torch.nn.functional.Bilinear, https://github.com/pytorch/pytorch/issues/48389
- [ ] add more modes to torch.nn.functional.grid_sample, https://github.com/pytorch/pytorch/issues/25039
- [ ] review torch.nn.functional.grid_sample documentation, https://github.com/pytorch/pytorch/issues/5565

**Bugs:**

- [ ] torch.nn.functional.grid_sample downsampling doesn't always match torch.nn.functional.interpolate downsampling, https://github.com/pytorch/pytorch/issues/21457
- [ ] torch.nn.functional.interpolate behaves unexpectedly with output size 1, https://github.com/pytorch/pytorch/issues/30565

**Comment by rgommers** [comment]
Note that `scipy.interpolate` has two sets of interpolation functionality: a functional style one (`interp1d`, `interp2d`, `spl*`, etc.) and a class-based one (`InterpolatedUnivariateSpline` et al.). The former is legacy, the latter is preferred.

`scipy.ndimage` indeed also has a lot of interpolation functionality. One of the most powerful functions is `ndimage.map_coordinates`, and for regular grids it's often preferred over what's in `scipy.interpolate`. Worth adding to this list of functions to consider I think.

**Comment by heitorschueroff** [comment]
cc @fmassa who's done some work in interpolations

**Comment by 0x00b1** [comment]
A friendly bump for what I consider an important missing feature.

@rgommers How do you feel about the class-based design in `scipy.interpolate`? I would be especially interested in reading about the trade-offs you and the other authors (e.g., @pv) made at the time and whether you’d still make the same decisions today. 

I believe the limited use of classes in the PyTorch API was one reason for its success (i.e., composition feels flexible yet natural and it’s relatively easy for users to discover the API). It also makes me feel like the limited use of classes in the design signals to users, when they encounter a class (e.g., Distribution), to be mindful of class-specific features (instead of just assuming classes were used for strictly namespace or organizational reasons). Is this a domain you think truly benefits from class features?

**Comment by rgommers** [comment]
> @rgommers How do you feel about the class-based design in `scipy.interpolate`? I would be especially interested in reading about the trade-offs you and the other authors (e.g., @pv) made at the time and whether you’d still make the same decisions today.

I don't think I was involved in the initial design, both APIs are from before 2010.

> I, personally, believe the limited use of classes in the PyTorch API was one reason for its success (i.e., composition feels flexible yet natural and it’s relatively easy for users to discover the API). It also makes me feel like the limited use of classes in the design signals to users, when they encounter a class (e.g., Distribution), to be mindful of class-specific features (instead of just assuming classes were used for strictly namespace or organizational reasons). Is this a domain you think truly benefits from class features?

I tend to agree. My preferred APIs - `griddata` for unstructured grids, and `map_coordinates` for structured grids - are functional. The class-based API (`UnivariateSpline` et al.) makes sense when you want to give users access to spline representations, allow integrating splines, taking derivates, etc., so I think that is a better fit for splines on regular grids than the old-style `splprep` et al. Classes also have the advantage of separating the construction of the interpolator from evaluating at different coordinates - making `griddata` do caching when you really need performance is a pain.

But for PyTorch that's too much functionality I'd think. Probably providing access to spline representations and integrating over curves should be out of scope. A set of powerful functions like `map_coordinates` and `griddata` would fit well in PyTorch core though.

**Comment by amitabe** [comment]
Is there any progress on these issues?

**Comment by fzimmermann89** [comment]
Any news? IMHO at least some 1D linear interpolation should be part of basic torch..

**Comment by sehoffmann** [comment]
Looking forward for this.
There is already this 3rd party implementation for 1d interpolation out there: https://github.com/aliutkus/torchinterp1d

**Comment by ferrarioa5** [comment]
I wrote a pytorch C++/CUDA extension library a RegularGridInterpolation analogous to Scipy [here](https://github.com/ferrarioa5/pytorch_interpolation). For now, only zero-padding and interpolation on a regular 2d grid are implemented.
